Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+                    guard nextTask.readyTime <= newTime else {
+                        break
+                    }
+
+                    // Now we want to grab all tasks that are ready to execute at the same
+                    // time as the first.
+                    var tasks = Array<EmbeddedScheduledTask>()
+                    while let candidateTask = self.scheduledTasks.peek(), candidateTask.readyTime == nextTask.readyTime {
+                        tasks.append(candidateTask)
+                        self.scheduledTasks.pop()
+                    }
+
+                    // Set the time correctly before we call into user code, then
+                    // call in for all tasks.
+                    self._now.store(nextTask.readyTime.uptimeNanoseconds)
+
+                    for task in tasks {
+                        task.task()
+                    }
+                }
+
+                // Finally ensure we got the time right.
+                self._now.store(newTime.uptimeNanoseconds)
+
+                continuation.resume()
+            }
+        }
+    }
+
+    /// Executes the given function in the context of this actor. This is useful when it's necessary to be confident that an operation
+    /// is "blocking" the actor. As long as you are executing, nothing else can execute in this loop.
+    public func executeInContext<ReturnType: Sendable>(_ task: @escaping @Sendable () throws -> ReturnType) async throws -> ReturnType {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<ReturnType, Error>) in
+            self.queue.async {
+                do {
+                    continuation.resume(returning: try task())
+                } catch {
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
+    }
+
+    internal func drainScheduledTasksByRunningAllCurrentlyScheduledTasks() {
+        var currentlyScheduledTasks = self.scheduledTasks
+        while let nextTask = currentlyScheduledTasks.pop() {
+            self._now.store(nextTask.readyTime.uptimeNanoseconds)
+            nextTask.task()
+        }
+        // Just fail all the remaining scheduled tasks. Despite having run all the tasks that were
+        // scheduled when we entered the method this may still contain tasks as running the tasks
+        // may have enqueued more tasks.
+        while let task = self.scheduledTasks.pop() {
+            task.fail(EventLoopError.shutdown)
+        }
+    }
+
+    private func _shutdownGracefully() {
+        dispatchPrecondition(condition: .onQueue(self.queue))
+        self.drainScheduledTasksByRunningAllCurrentlyScheduledTasks()
+    }
+
+    /// - see: `EventLoop.shutdownGracefully`
+    public func shutdownGracefully(queue: DispatchQueue, _ callback: @escaping (Error?) -> Void) {
+        self.queue.async {
+            self._shutdownGracefully()
+            queue.async {
+                callback(nil)
+            }
+        }
+    }
```

Comment:
Take it or leave it.
```suggestion
    public func shutdownGracefully(queue callbackQueue: DispatchQueue, _ callback: @escaping (Error?) -> Void) {
        self.queue.async {
            self._shutdownGracefully()
            callbackQueue.async {
                callback(nil)
            }
        }
    }
```

Reply:
Leave it: this is part of the `EventLoop` protocol.

Reply:
I don't think this would be a problem though, we just name the parameter differently. But I'm fine with leaving it.